python/ray/workflow/tests/test_http_events.py

```diff
@@ -0,0 +1,142 @@
+
+    assert event_msg == 'event_message'
+
+    serve.shutdown()
+
+@pytest.mark.parametrize(
+    "workflow_start_regular_shared",
+    [
+        {
+            "num_cpus": 4,  # TODO (Alex): When we switch to the efficient event
+            # implementation we shouldn't need these extra cpus.
+        }
+    ],
+    indirect=True,
+)
+def test_dynamic_event_by_http(workflow_start_regular_shared):
+    """If a workflow has dynamically generated event arguments, it should return the event as if the event
+    was declared statically.
+    """
+    serve.start(detached=True)
+    HTTPEventProvider.deploy()
+
+    @ray.remote
+    def send_event():
+        from time import sleep
+        sleep(1)
+        resp = requests.post('http://127.0.0.1:8000/event/send_event/workflow_id', \
+            json={'workflow_id':'workflow_id','event_key':'event_key','event_payload':'event_message_dynamic'})
+
+    @ray.remote
+    def return_dynamically_generated_event():
+        event_step = workflow.wait_for_event(HTTPListener, workflow_id="workflow_id", event_key="event_key")
+        return workflow.continuation(event_step)
+
+    workflow.create(return_dynamically_generated_event.bind()).run_async("workflow_test_dynamic_event_by_http")
+
+    ray.get(send_event.remote())
+    event_msg = ray.get(workflow.get_output("workflow_test_dynamic_event_by_http"))
+
+    assert event_msg == 'event_message_dynamic'
+
+    serve.shutdown()
+
+@pytest.mark.parametrize(
+    "workflow_start_regular_shared",
+    [
+        {
+            "num_cpus": 4,  # TODO (Alex): When we switch to the efficient event
+            # implementation we shouldn't need these extra cpus.
+        }
+    ],
+    indirect=True,
+)
+def test_multiple_events_by_http(workflow_start_regular_shared):
+    """If a workflow has multiple event arguments, it should wait for them at the
+    same time.
+    """
+    serve.start(detached=True)
+    HTTPEventProvider.deploy()
+
+    @ray.remote
+    def send_event1():
+        from time import sleep
+        sleep(3)
+        resp = requests.post('http://127.0.0.1:8000/event/send_event/wid', \
+            json={'workflow_id':'wid','event_key':'e1','event_payload':'hello'})
+
+    @ray.remote
+    def send_event2():
+        from time import sleep
+        sleep(4)
+        resp = requests.post('http://127.0.0.1:8000/event/send_event/wid', \
+            json={'workflow_id':'wid','event_key':'e2','event_payload':'world'})
+
+    @ray.remote
+    def trivial_step(arg1, arg2):
+        return f"{arg1} {arg2}"
+
+    event1_promise = workflow.wait_for_event(HTTPListener, workflow_id="wid", event_key="e1")
+    event2_promise = workflow.wait_for_event(HTTPListener, workflow_id="wid", event_key="e2")
+    promise = workflow.create(
+        trivial_step.bind(event1_promise, event2_promise)).run_async("workflow_test_multiple_event_by_http")
+
+    ray.get(send_event1.remote())
+    ray.get(send_event2.remote())
```

Comment:
Any reason using remote function here? Why not just send post here directly? Also why do we sleep inside remote function? Is there any race condition which we are not fixing?

Reply:
If it hangs, why? Because from my understanding, it shouldn't hangs.

No specific reason not using this. But we need the reason why we need to use this.

Reply:
remote removed after updating the tests to loop over 404 responses.